File: src/main/java/net/lomeli/lomlib/client/models/ModelHandler.java
package net.lomeli.lomlib.client.models;

import com.google.common.collect.Lists;

import java.util.List;

import net.lomeli.lomlib.LomLib;
import net.minecraft.block.Block;
import net.minecraft.client.Minecraft;
import net.minecraft.client.renderer.ItemMeshDefinition;
import net.minecraft.client.renderer.block.model.ModelResourceLocation;
import net.minecraft.client.renderer.color.ItemColors;
import net.minecraft.item.Item;

import net.minecraftforge.client.model.ModelLoader;
import net.minecraftforge.fml.common.Loader;
import net.minecraftforge.fml.common.LoaderState;

public class ModelHandler {

    private static List<IColorProvider> colorProviders = Lists.newArrayList();

    public static void registerModel(IModelHolder holder) {
        if (holder == null)
            return;
        if (Loader.instance().getModState(Loader.instance().activeModContainer()) != LoaderState.ModState.PREINITIALIZED) {
            LomLib.logger.logError("Must be registered in Pre-Init");
        }
        if (holder instanceof IColorProvider)
            colorProviders.add((IColorProvider) holder);
        if (holder instanceof IMeshVariant) {
            ItemMeshDefinition def = ((IMeshVariant) holder).getCustomMesh();
            if (def != null) {
                ModelLoader.setCustomMeshDefinition(holder instanceof Item ? (Item) holder : holder instanceof Block ? Item.getItemFromBlock((Block) holder) : null, def);
                return;
            }
        }

        if (holder instanceof Item) {
            Item item = (Item) holder;
            String[] variants = holder.getVariants();
            if (variants != null && variants.length > 0) {
                for (int i = 0; i < variants.length; i++) {
                    String name = variants[i];
                    ModelResourceLocation loc = new ModelResourceLocation(name, "inventory");
                    ModelLoader.setCustomModelResourceLocation(item, i, loc);
                }
            }
        } else if (holder instanceof Block) {
            Block block = (Block) holder;
            String[] variants = holder.getVariants();
            if (variants != null && variants.length > 0) {
                for (int i = 0; i < variants.length; i++) {
                    String name = variants[i];
                    ModelResourceLocation loc = new ModelResourceLocation(name, "inventory");
                    if (holder instanceof IModelVariant) {
                        loc = new ModelResourceLocation(name, ((IModelVariant) holder).getModelTypes()[i]);
                        ModelLoader.registerItemVariants(Item.getItemFromBlock(block), loc);
                    }
                    ModelLoader.setCustomModelResourceLocation(Item.getItemFromBlock(block), i, loc);
                }
            }
        }
    }

    public static void registerModel(Object obj) {
        if (obj instanceof IModelHolder) registerModel((IModelHolder) obj);
    }

    public static void registerColorProviers() {
        ItemColors colors = Minecraft.getMinecraft().getItemColors();
        colorProviders.stream().filter(c -> c != null).forEach(c -> colors.registerItemColorHandler(c.getColor(), (Item) c));
    }
}

File: src/main/kotlin/net/lomeli/lomlib/client/models/ModelHandler.kt
package net.lomeli.lomlib.client.models

import com.google.common.collect.Lists
import net.lomeli.lomlib.LomLib
import net.minecraft.block.Block
import net.minecraft.client.Minecraft
import net.minecraft.client.renderer.block.model.ModelResourceLocation
import net.minecraft.item.Item
import net.minecraftforge.client.model.ModelLoader
import net.minecraftforge.fml.common.Loader
import net.minecraftforge.fml.common.LoaderState

object ModelHandler {
    private val colorProviders = Lists.newArrayList<IColorProvider>()

    fun registerModel(holder: IModelHolder) {
        if (holder == null)
            return
        if (Loader.instance().loaderState != LoaderState.PREINITIALIZATION)
            LomLib.logger.logError("Must be registered in Pre-Init")
        if (holder is IColorProvider)
            colorProviders.add(holder)
        if (holder is IMeshVariant) {
            val def = holder.getCustomMesh()
            if (def != null) {
                ModelLoader.setCustomMeshDefinition(if (holder is Item) holder else if (holder is Block) Item.getItemFromBlock((holder as Block?)!!) else null, def)
                return
            }
        }

        if (holder is Item) {
            val variants = holder.getVariants()
            if (variants != null && variants.size > 0) {
                for (i in variants.indices) {
                    val name = variants[i]
                    val loc = ModelResourceLocation(name, "inventory")
                    ModelLoader.setCustomModelResourceLocation(holder, i, loc)
                }
            }
        } else if (holder is Block) {
            val variants = holder.getVariants()
            if (variants != null && variants.size > 0) {
                for (i in variants.indices) {
                    val name = variants[i]
                    var loc = ModelResourceLocation(name, "inventory")
                    if (holder is IModelVariant) {
                        loc = ModelResourceLocation(name, holder.getModelTypes()[i])
                        ModelLoader.registerItemVariants(Item.getItemFromBlock(holder)!!, loc)
                    }
                    ModelLoader.setCustomModelResourceLocation(Item.getItemFromBlock(holder)!!, i, loc)
                }
            }
        }
    }

    fun registerModel(obj: Any) {
        if (obj is IModelHolder) registerModel(obj)
    }

    fun registerColorProviders() {
        val colors = Minecraft.getMinecraft().itemColors
        for (c in colorProviders) {
            if (c != null) colors.registerItemColorHandler(c.getColor(), c as Item)
        }
    }
}